components/builder-depot/src/backend/s3.rs

```diff
@@ -169,45 +193,53 @@ impl S3Handler {
         }
     }
 
-    fn single_upload(&self, key: String, hart: File, path_attr: &str) -> Result<()> {
+    fn single_upload(&self, key: &str, hart: File, path_attr: &PathBuf) -> Result<()> {
         Counter::SingleUploadRequests.increment();
         let start_time = PreciseTime::now();
         let mut reader = BufReader::new(hart);
         let mut object: Vec<u8> = Vec::new();
         let bucket = self.bucket.clone();
-        let _complete = reader.read_to_end(&mut object)?;
+        let _complete = reader.read_to_end(&mut object).map_err(Error::IO);
 
         let mut request = PutObjectRequest::default();
-        request.key = key;
+        request.key = key.to_string();
         request.bucket = bucket;
         request.body = Some(object);
 
-        match self.client.put_object(&request).sync() {
-            Ok(_) => {
-                let end_time = PreciseTime::now();
-                info!(
-                    "Upload completed for {} (in {} sec):",
-                    &path_attr,
-                    start_time.to(end_time)
-                );
-                Ok(())
-            }
-            Err(e) => {
-                Counter::UploadFailures.increment();
-                warn!("Upload failed for {}: ({:?})", &path_attr, e);
-                Err(Error::PackageUpload(e))
+        let fqpi = path_attr.clone().into_os_string().into_string().unwrap();
+
+        let put_future = self.client.put_object(&request).then(move |result| {
```

Comment:
Is the change to use futures required?

Reply:
It's using futures either way, this just moves us closer to having our uploads handled asynchronously. If you'd prefer we keep `sync`ing it should be trivial to back this bit out.

Reply:
Since this is the only future we are waiting on when we run the reactor they are both doing essentially the same thing.

Reply:
It's not the only thing we're waiting on, we've got a whole bunch of stuff in this module that uses sync this is just the first couple things I've removed the sync from. Do you think it's better to just phone it in and remove this change?